### PR TITLE
Support for multiple contracts per account

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/dapperlabs/flow-emulator/issues/98
moved from: https://github.com/dapperlabs/flow-go/pull/4862

### CODE REVIEW

File: fvm/env.go
Comment (janezpodhostnik) [inline fvm/env.go:40]: @turbolent this came with the new version of cadence. I hope I implemented it as intended.

File: go.mod
Comment (janezpodhostnik) [inline go.mod:59]: @turbolent I have to add this otherwise it doesn't build. Is this ok?

Comment (turbolent) [reply go.mod:59]: Ah, right. The replace is only local to the package itself (i.e. onflow/cadence), and won't be applied in packages importing it (like here, flow-go). Let me look into this

File: fvm/env.go
Comment (turbolent) [reply fvm/env.go:40]: Yep, this looks great!

File: fvm/env.go
Comment (turbolent) [inline fvm/env.go:46]: Minor:
```suggestion
	return hasher.ComputeHash(data)
```

File: fvm/env.go
Comment (turbolent) [inline fvm/env.go:400]:
```suggestion
		panic("RemoveAccountContractCode is not supported by this environment")
```

File: go.mod
Comment (turbolent) [reply go.mod:59]: Asked the rest of the team here: https://axiomzen.slack.com/archives/CEXKUH74G/p1601594114204500

File: fvm/state/accounts.go
Comment (turbolent) [inline fvm/state/accounts.go:313]: The first two arguments, "controller" and "owner", are historical and soon will be merged into just one. 

Instead of using one of them to store the contract name, keep the first two values as is, but store under a different key instead, e.g. `fmt.Sprintf("%s.%s", keyCode, name)`.

This also needs to be changed for all the other `Get`/`Set` cases below

File: engine/verification/test/verification_test.go
Comment (janezpodhostnik) [inline engine/verification/test/verification_test.go:153]: Tests keep failing if this is too low. I need to check whats up with that before merging!

**Comment by turbolent** [comment]
@janezpodhostnik Looks great! Can you please target this against a feature branch, we can't merge this to master yet.

**Comment by janezpodhostnik** [comment]
> @janezpodhostnik Looks great! Can you please target this against a feature branch, we can't merge this to master yet.

done

**Review by turbolent — APPROVED** [review]
Looks good, awesome work @janezpodhostnik!

Would be great to get a second review from someone more familiar with FVM, @psiemens @m4ksio could you maybe have a look please?

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:351]: Do you think we should store each contract in its own register slot?

RLP was a solution we used for account keys before we had given sufficient thought to how storage should work. I don't think we have to use it here unless there's a compelling need

**Review by psiemens — COMMENTED** [review]
This is a huge feature, great job implementing it!

I left a few comments about the design, my biggest being a question about the usage of RLP

File: fvm/state/accounts.go
Comment (turbolent) [reply fvm/state/accounts.go:351]: The code of each contract is stored in a separate register – this "contracts" value / RLP list is the set of all contract names. 

@janezpodhostnik I was also confused about the naming. Some ideas to alleviate that:

- Maybe rename `ContractList` to `ContractNames`
- Maybe rename `Accounts.GetContracts` to `GetContractNames` (https://github.com/onflow/flow-go/pull/7/files#diff-177914ac132c7b58a325a8cd8eda6faf6d515c16c11a95ca99fe24a0e2dfff64R384). `GetContract` returns the contract code, but the current plural (`GetContracts`) only returns the names.
- Rename the variables in the function similarly from "contracts" to "contractNames", e.g. `newContracts` to `newContractNames`, `prevContracts` to `prevContractNames`, etc.

File: fvm/state/accounts.go
Comment (janezpodhostnik) [reply fvm/state/accounts.go]: Had to add another method so I don't return `ContractsList` from a public function

File: fvm/state/accounts.go
Comment (janezpodhostnik) [reply fvm/state/accounts.go]: Renaming was a good idea to make it clearer whats actually stored. I also switched to cbor encoder/decoder, or did you have something else in mind @psiemens?

Comment (turbolent) [reply fvm/state/accounts.go]: @janezpodhostnik Renaming looks good! 

I don't know what's commonly used for serialization on the EN/FVM side, might be good to keep it consistent.

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:363]:
```suggestion
		return fmt.Errorf("cannot retrieve current contract names: %w", err)
```

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:301]:
```suggestion
func (a *Accounts) getContract(name string, address flow.Address) ([]byte, error) {
```

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:313]:
```suggestion
func (a *Accounts) setContract(name string, address flow.Address, code []byte) error {
```

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:61]: Can you move this code to the bottom of the file? Generally we try to keep the main constructor as the first entry, followed by receiver functions

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:393]: I think it'd be fine to just keep the single `GetContractNames` function

File: fvm/state/accounts_test.go
Comment (psiemens) [inline fvm/state/accounts_test.go:81]: You can use `require.NoError(t, err)` in these places

File: fvm/state/accounts_test.go
Comment (psiemens) [inline fvm/state/accounts_test.go:88]:
```suggestion
                require.Len(t, contracts, 1, "There should only be one contract")
                require.Equal(t, "Dummy", contracts[0])
```

File: fvm/state/accounts.go
Comment (psiemens) [inline fvm/state/accounts.go:106]: Is the `Code` field still needed on this model?

File: fvm/state/accounts.go
Comment (psiemens) [reply fvm/state/accounts.go]: Ah, I see! I was a bit confused before.

Yep, CBOR is a good choice.

I suppose it would be even better if we could do a prefix scan for all of the `contract_` keys. Is that why you brought it up on Slack, @turbolent? It looks like @ramtinms and @m4ksio are planning on adding support for that. How hard do you think it would be to remove this list in the future?

**Review by psiemens — APPROVED** [review]
Looks good, just left a few more comments but nothing related to the logic.

File: fvm/state/accounts.go
Comment (janezpodhostnik) [reply fvm/state/accounts.go]: The hard part about switching this to prefixes would be the register migration. The code change would be very minor.

File: fvm/state/accounts.go
Comment (m4ksio) [inline fvm/state/accounts.go:423]: Maybe I miss something, but do we use this list for anything else than checking membership?

Comment (m4ksio) [inline fvm/state/accounts.go:423]: Do I miss something, or we need this list only to check for presence?

Comment (janezpodhostnik) [reply fvm/state/accounts.go:423]: We also use it in env.go ResolveLocation. If in cadence an import is used as:
```
import from A
// instead of 
// import foo from A
```

That needs to resolve to all contracts on A, which we can only get by keeping a list of contracts on A (or by prefix search when that will be available)

Comment (janezpodhostnik) [reply fvm/state/accounts.go:423]: same reply as here: https://github.com/onflow/flow-go/pull/7#discussion_r508674384

File: fvm/state/accounts.go
Comment (turbolent) [reply fvm/state/accounts.go]: @psiemens I had asked about the prefix scan on Slack because we need it for general iteration over all storage paths of an account, but yeah, it could be used here as well.

File: fvm/state/accounts.go
Comment (m4ksio) [reply fvm/state/accounts.go:423]: sure - does this require it to be sorted though?

Comment (turbolent) [reply fvm/state/accounts.go:423]: Yes, it needs to be sorted to be deterministic. If it is not, the register reads/touches will not be consistent across different machines

Comment (janezpodhostnik) [reply fvm/state/accounts.go:423]: I added the sorting to have a deterministic state commitment. The sorting wont be necessary if we just get them by key prefix.

**Comment by turbolent** [comment]
@janezpodhostnik looks like there are enough approvals now, feel free to merge this and then https://github.com/dapperlabs/flow-emulator/pull/105 can be updated to use this feature branch

**Pull request merged**
